Print pointer register names only when they are requested

dylanmckay opened the pull request:
This PR changes `AVRRegisterInfo.td` to treat `X`, `Y`, and `Z` as aliases of `r27:r26`, `r29:r28`, and `r31:r30` respectively.

It modifies TableGen to match based on register aliases as well as their names in `AVRAsmParser::MatchRegisterName`.

It also modifies the instruction printer to only print the `X`, `Y`, and `Z` aliases if the instruction has operands of type `PTRREGS`, `PTRDISPREGS`, or `ZREGS`. This causes `movw r30, r28` to be printed unchanged - `movw r30, r28`, instead of `movw Z, Y`.

@agnat Could you look over this?

agnat commented:
lol. I was working on the same patch.

> Could you look over this?

Will do.

agnat commented on lib/Target/AVR/InstPrinter/AVRInstPrinter.cpp:
```diff
@@ -76,15 +78,24 @@ void AVRInstPrinter::printInst(const MCInst *MI, raw_ostream &O,
+      RegToPrint = RegNum; // print the entire register pair
+    else
+      RegToPrint = RegLoNum; // print the lower register
   }
```
The whole if block is odd... Let's rewrite the condition without magic numbers. We get:

````
if (MRI.getSubReg(RegNum, AVR::sub_lo) != AVR::NoRegister)
````

Now it's easy to see that it checks the same thing twice. Let's rewrite the whole thing as:

````
  auto RegLoNum = MRI.getSubReg(RegNum, AVR::sub_lo);
  return getRegisterName(RegLoNum != AVR::NoRegister ? RegLoNum : RegNum);
````

... or some such.

dylanmckay replied:
Good point. This is fixed.

agnat commented on utils/TableGen/AsmMatcherEmitter.cpp:
```diff
@@ -2187,6 +2187,13 @@ static void emitMatchRegisterName(CodeGenTarget &Target, Record *AsmParser,
+    for(auto AltName : AltNames) {
+        Matches.emplace_back(AltName,
+                             "return " + utostr(Reg.EnumValue) + ";");
+    }
```
Not sure about this. It is a change in behavior for other backends using `RegAltNames`. Shouldn't that be optional, somehow?

dylanmckay replied:
I thought it wouldn't, but `AArch64` fails to build. I'm investigating now.

agnat commented:
I finished my review. LGTM.

Oh, and since "register aliases" are ... something else, could you use "alt names" or the like in the final commit message?

agnat commented:
Btw, I came across this [little bug](https://github.com/agnat/llvm/commit/8beb771ca951c870ff5d086ac7cbd8658e5fd700) regarding alt names. Do you have any experience in pushing patches upstream?

dylanmckay commented:
I don't, but I am under the impression that you attach a patch, describing it in an email, and send it to the `llvm-commits` list.

dylanmckay commented:
I changed the behaviour from extending `MatchRegisterName` to adding `MatchRegisterAltName`, and renamed everything from "register alias" to "alternative name".

I also sent the target-agnostic part of the patch upstream (see the mailing list post [here](http://lists.cs.uiuc.edu/pipermail/llvm-commits/Week-of-Mon-20150615/281983.html)).

Committed in b21056bae1f11906de38cbdc301525174291a7ac.
